IdentityCore/src/throttling/MSIDThrottlingService.m

```diff
@@ -0,0 +1,471 @@
+    throttleType = [self isResponseUIRequiredThrottleType:errorResponse];
+    return throttleType;
+}
+
+/**
+ 429 throttle conditions:
+ - HTTP Response code is 429 or in 5xx range
+ - OR Retry-After in response header
+ */
+- (MSIDThrottlingType)isResponse429ThrottleTypeWithErrorResponse:(NSError * _Nullable )errorResponse
+                                                 isSSOExtRequest:(BOOL)isSSOExtRequest
+                                                           error:(NSError *_Nullable *_Nullable)error
+{
+    MSIDThrottlingType throttleType = MSIDThrottlingTypeNone;
+    // TODO: Let assume the error here is MSIDError, we will deal with conversion later:
+    // In both scenarios (server error or SSO-ext internal error) broker creates MSALError and return to calling app
+    
+    NSString *httpResponseCode = errorResponse.userInfo[MSIDHTTPResponseCodeKey];
+    NSInteger responseCode = [httpResponseCode intValue];
+    if (responseCode == 429) throttleType = MSIDThrottlingType429;
+    if (responseCode >= 500 && responseCode <= 599) throttleType = MSIDThrottlingType429;
+    NSDate *retryHeaderDate = [self getRetryDateFromErrorResponse:errorResponse];
+    if (retryHeaderDate)
+    {
+        throttleType = MSIDThrottlingType429;
+    }
+    
+    
+    return throttleType;
+}
+
+- (void)updateThrottlingDatabaseWithRequest:(id<MSIDThumbprintCalculatable> _Nonnull)request
+                              errorResponse:(NSError * _Nullable )errorResponse
+                            isSSOExtRequest:(BOOL)isSSOExtRequest
+                                returnError:(NSError *_Nullable *_Nullable)error
+{
+    NSError *localError = nil;
+    MSIDThrottlingType throttleType = [self getThrottleTypeFrom:request
+                                                  errorResponse:errorResponse
+                                                isSSOExtRequest:isSSOExtRequest
+                                                          error:&localError] ;
+    if (localError)
+    {
+        MSID_LOG_WITH_CTX(MSIDLogLevelError, self.context, BASE_MSG_UPDATING_ERROR, localError);
+        if (error)
+        {
+            *error = localError;
+        }
+        return;
+    }
+    
+    if (throttleType == MSIDThrottlingTypeNone) return;
+    
+    // create throttling record and update to db
+    [self createDBRecordAndUpdateWithRequest:request
+                               errorResponse:errorResponse
+                             isSSOExtRequest:isSSOExtRequest
+                                throttleType:throttleType
+                                 returnError:error];
+    
+    return;
+}
+
+
+/**
+ * If not 429, we check if is appliable for UIRequired:
+ * error response can be: invalid_request, invalid_client, invalid_scope, invalid_grant, unauthorized_client, interaction_required, access_denied
+ */
+- (MSIDThrottlingType)isResponseUIRequiredThrottleType:(NSError *)errorResponse
```

Comment:
I think I got confused by the name of the method before since it accounts for more than just UI required errors :)

Reply:
I'm struggling with the name of throttle types :D. Currently, the 2 names are 429Throttle and UIRequiredThrottle. UIRequiredThrottle -> InteractionRequiredThrottle sounds better?